src/libraries/System.Net.Quic/tests/FunctionalTests/QuicStreamTests.cs

```diff
@@ -228,6 +228,84 @@ public async Task GetStreamIdWithoutStartWorks()
             Assert.Equal(0, clientStream.StreamId);
         }
 
+        [Fact]
+        public async Task LargeDataSentAndReceived()
+        {
+            byte[] data = Enumerable.Range(0, 64 * 1024).Select(x => (byte)x).ToArray();
+            const int NumberOfWrites = 1024;       // total sent = 64M
+
+            using QuicListener listener = CreateQuicListener();
+
+            for (int j = 0; j < 100; j++)
+            {
+                Task listenTask = Task.Run(async () =>
+                {
+                    using QuicConnection connection = await listener.AcceptConnectionAsync();
+                    await using QuicStream stream = await connection.AcceptStreamAsync();
+                    byte[] buffer = new byte[data.Length];
+
+                    for (int i = 0; i < NumberOfWrites; i++)
+                    {
+                        int totalBytesRead = 0;
+                        while (totalBytesRead < data.Length)
+                        {
+                            int bytesRead = await stream.ReadAsync(buffer.AsMemory().Slice(totalBytesRead));
+                            Assert.NotEqual(0, bytesRead);
+                            totalBytesRead += bytesRead;
+                        }
+
+                        Assert.Equal(data.Length, totalBytesRead);
+                        Assert.True(data.AsSpan().SequenceEqual(buffer));
+                    }
+
+                    for (int i = 0; i < NumberOfWrites; i++)
+                    {
+                        await stream.WriteAsync(data);
+                    }
+
+                    await stream.WriteAsync(Memory<byte>.Empty, endStream: true);
+
+                    await connection.CloseAsync(errorCode: 0);
```

Comment:
You need to wait for stream shutdown before closing the connection, or the stream data is not guaranteed to get there.

Reply:
Thanks. Presumably if I Dispose the stream, I get the guarantee?

Reply:
Alternatively, if I just remove the call to connection.CloseAsync and let it be disposed by the using statement, will that guarantee that the data is sent as well?

I'm trying to understand how customers are expected to use this.

Reply:
Not in current implementation. My proposal in https://github.com/dotnet/runtime/issues/756#issuecomment-697979873 would make that the case.

Reply:
ah, right.